plutus-core/docs/equalsCanonical.md

```diff
@@ -0,0 +1,203 @@
+
+```
+mkCanonical : all a. (a -> data) -> a -> canonical a
+```
+
+This builtin says that as long as any value of type `a` can be converted to a `data`, it's safe to say the value is going to be in canonical form when we reach it during evaluation. Here's its implementation:
+
+```haskell
+    toBuiltinMeaning _ver MkCanonical =
+        makeBuiltinMeaning
+            mkCanonicalPlc
+            (\_ _ _ -> ExBudgetLast mempty)
+      where
+        mkCanonicalPlc :: Opaque val (a -> Data) -> Opaque val a -> Opaque val (Canonical a)
+        mkCanonicalPlc _ = coerce
+```
+
+Now this can of course be problematic if the user supplies an erroring `a -> data` function as a "proof", but given that `error` isn't reflected in the type system, we can't do better than [Fast and loose reasoning is morally correct](https://www.researchgate.net/publication/220997017_Fast_and_loose_reasoning_is_morally_correct) anyway -- implicit side effects are like that.
+
+What is great about this builtin is that it's completely erasable, since its implementation ignores the first argument and returns the second one without changing it in any way on the Plutus side (even on the Haskell side that `coerce` doesn't survive compilation). I.e. it's basically the same as `trace` except it doesn't even have any side effect and we already know how to erase `trace`. Thus whatever we provide as that `a -> Data` function can be erased without pointlessly wasting precious bytes of script size.
+
+Moreover, constract writers who don't care about types (which is often the case) can directly replace `equalsData` with `equalsCanonical` and still have everything working, since `canonical` isn't an actual built-in type whose values go into the `Constant` constructor, it's merely a [ghost of a departed proof](https://kataskeue.com/gdp.pdf).
+
+How should we implement `Canonical` then given that it's a ghost in Plutus and only a type-level entity in Haskell not needing any term-level content? Well, the question kinda answers itself:
+
+```haskell
+type Canonical :: GHC.Type -> GHC.Type
+data Canonical a
+```
+
+Moreover, once the grand universe split is complete (see [this](https://input-output.atlassian.net/browse/PLT-810) internal Jira ticket), we won't even need to add a runtime type tag for `Canonical` to the term-level universe, limiting its existence only to the type-level universe.
+
+Long story short, `Canonical` and `canonical` are merely for ensuring that the type checker still prevents nonsensical programs from being created, they don't affect evaluation time (as long as `mkCanonical` invocations are properly erased) and are entirely opt-in (i.e. people who don't care about type safety don't need to know anything about the `canonical` type and can use `equalsCanonical` directly).
+
```

Comment:
If what we are interested in is things that can be compared, why bother with `canonical` and not have an `Eq` type constructor, and choose just the types that we want?
```
mkInteger : integer -> Eq Integer
mkPair : all a b. Eq a -> Eq b -> Eq (pair a b)
mkList : all a. list (Eq a) -> Eq (list a)
...
```

Reply:
Your examples are all built-in types, how would you choose all appropriate SOPs via built-in functions in addition to those? What types would such builtins even have?

Reply:
For SOPs, it can't be a builtin, since builtins can't handle SOPs. There are different alternatives. Perhaps the closest one to the proposed one above would need a language construct:
`mkSOP : sop [(Eq a) (Eq b) ] ... [  Eq c] ->  Eq (sop [a b] ... [c])` 
(I don't quite remember the concrete syntax for SOP types)
Note the informal `...`. These are the same as those in the description of the `constr` type.

Another alternative would be a `constrEq` construct that behaves like `constr` but only accepts `Eq` types.

Another alternative is to have implicit castings, by modifying the typing rules. 

There is a whole range of possible solutions, each with their own advantages and disadvantages.

Reply:
> For SOPs, it can't be a builtin, since builtins can't handle SOPs.

Then it can't be fast, which defeats the purpose.

> There is a whole range of possible solutions,

If you're going to propose any of those, could you do that in the form of the "A bird's-eye view" section of the updated proposal?